File: objc/memento/Memento.m
#import "Memento.h"

@implementation EditorMemento

- (instancetype)initWithContent:(NSString *)content {
    self = [super init];
    if (self) {
        _content = [content copy];
    }
    return self;
}

@end

@implementation TextEditor

- (EditorMemento *)save {
    return [[EditorMemento alloc] initWithContent:self.content];
}

- (void)restore:(EditorMemento *)memento {
    self.content = memento.content;
}

@end

@implementation History {
    NSMutableArray<EditorMemento *> *_snapshots;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _snapshots = [NSMutableArray array];
    }
    return self;
}

- (void)push:(EditorMemento *)memento {
    [_snapshots addObject:memento];
}

- (nullable EditorMemento *)pop {
    if (_snapshots.count == 0) {
        return nil;
    }
    EditorMemento *last = _snapshots.lastObject;
    [_snapshots removeLastObject];
    return last;
}

- (NSUInteger)count {
    return _snapshots.count;
}

@end

File: java/memento/Memento.java
/**
 * 备忘录标记接口（Memento）：故意不暴露任何方法。
 * 管理者（History）只能持有这个引用来保存/传递快照，无法查看或修改其中的状态，
 * 只有发起人（TextEditor）自己知道如何解读真正的数据，从而保护了封装性。
 */
public interface Memento {
}
